core/src/main/java/org/apache/iceberg/rest/HTTPClient.java

```diff
@@ -0,0 +1,282 @@
+      return EntityUtils.toString(response.getEntity(), "UTF-8");
+    } catch (IOException | ParseException e) {
+      throw new RESTException(e, "Failed to convert HTTP response body to string");
+    }
+  }
+
+  // Per the spec, the only currently defined / used "success" responses are 200 and 202.
+  private static boolean isSuccessful(CloseableHttpResponse response) {
+    int code = response.getCode();
+    return code ==  HttpStatus.SC_OK || code == HttpStatus.SC_ACCEPTED;
+  }
+
+  private static ErrorResponse buildDefaultErrorResponse(CloseableHttpResponse response) {
+    String responseReason = response.getReasonPhrase();
+    String message =
+        responseReason != null && !responseReason.isEmpty() ? responseReason :
+            EnglishReasonPhraseCatalog.INSTANCE.getReason(response.getCode(), null /* ignored */);
+    String type = "RESTException";
+    return ErrorResponse.builder()
+        .responseCode(response.getCode())
+        .withMessage(message)
+        .withType(type)
+        .build();
+  }
+
+  private static void handleFailure(
+      CloseableHttpResponse response, String responseBody, Consumer<ErrorResponse> errorHandler) {
+    ErrorResponse errorResponse = null;
+
+    if (responseBody != null) {
+      try {
+        errorResponse = ErrorResponseParser.fromJson(responseBody);
+      } catch (Exception e) {
+        // It's possible to receive a non-successful response that isn't a properly defined ErrorResponse
+        // without any bugs in the server implementation. So we ignore this exception and build an error
+        // response for the user.
+        //
+        // For example, the connection could time out before every reaching the server, in which case we'll
+        // likely get a 5xx with the load balancers default 5xx response.
+        LOG.error("Failed to parse an error response. Will create one instead.", e);
+      }
+    }
+
+    if (errorResponse == null) {
+      errorResponse = buildDefaultErrorResponse(response);
+    }
+
+    errorHandler.accept(errorResponse);
+
+    // Throw a RESTException, in case the provided error handler does not throw.
+    throw new RESTException(
+        "Entered the error handler path but no exception was thrown for error response %s",
+        ErrorResponseParser.toJson(errorResponse));
+  }
+
+  /**
+   * Method to execute an HTTP request and process the corresponding response.
+   *
+   * @param method       - HTTP method, such as GET, POST, HEAD, etc.
+   * @param path         - URL path to send the request to
+   * @param requestBody  - Content to place in the request body
+   * @param responseType - Class of the Response type. Needs to have serializer registered with ObjectMapper
+   * @param errorHandler - Error handler delegated for HTTP responses which handles server error responses
+   * @param <T>          - Class type of the response for deserialization. Must be registered with the ObjectMapper.
```

Comment:
The object mapper is internal, so maybe a better way to state the requirement is that the class must be one of the responses in `org.apache.iceberg.rest.responses`? Same with `requests`.

Reply:
In another PR I was working on, I was going to add a marker interface for `RESTRequest` and `RESTResponse`.

Maybe a top level empty interface `RESTMessage` (similar to `HttpMessage` that many libraries use), and then have `RESTRequest` and `RESTResponse` stem from those.

I can submit that sooner so that we can add it in here?

Reply:
I opened a PR to add an interface for `RESTRequest` and `RESTResponse`: https://github.com/apache/iceberg/pull/4405

They can be attached as a follow up if that PR is merged later.

Reply:
That helps, but the caller still knows nothing about `ObjectMapper` so the Javadoc shouldn't mention it.